### PR TITLE
Fix Pod not being removed after istio-sidecar is removed 

### LINKED ISSUE #34482: KubernetesPodOperator shutting down istio sidecar but not deleting pod upon failure
### Apache Airflow version

2.7.1

### What happened

I start a simple pod (one container and that's it) with the KubernetesPodOperator from the cncf.kubernetes provider.

The task in my container fails (exit code non-zero).
The istio sidecar is indeed terminated but the pod itself remain in an `Error` status even if the `on_finish_action` parameter is set at `"delete_pod"` and the pod is never terminated.

It is expected that the pod is fully deleted, istio or not.

I found that there is a difference in treatment [upon deleting the pod](https://github.com/apache/airflow/blob/5b85442fdc19947e125dcb0591bd59a53626a27b/airflow/providers/cncf/kubernetes/operators/pod.py#L824) so it might be some specific case I'm not aware of.

I'll be happy to help either on documentation or fixing this small issue but I would need confirmation on what is the expected behavior.

### What you think should happen instead

I think the `on_finish_action="delete_pod"` should terminate the pod, not let it hanging the `Error` state with both containers stopped.

### How to reproduce

here is the simplest dag on how to reproduce on my end, note that istio is not visible here since managed cluster-wide.

```python3
from airflow import DAG
from airflow.utils.dates import days_ago
from airflow.providers.cncf.kubernetes.operators.kubernetes_pod import (
    KubernetesPodOperator,
)


with DAG(
    dag_id="issue_reproduction",
    schedule_interval=None,
    start_date=days_ago(0),
    is_paused_upon_creation=False,
    max_active_runs=1,
) as dag:
    pod = KubernetesPodOperator(
        task_id="issue_reproduction",
        image="ubuntu",
        cmds=["bash", "-cx"],
        arguments=["exit 1"],
        dag=dag,
        security_context={
            "runAsNonRoot": True,
            "runAsUser": 65534,
            "seccompProfile": {"type": "RuntimeDefault"},
        },
    )
```

### Operating System

local=macos, deployments=k8s

### Versions of Apache Airflow Providers

apache-airflow-providers-cncf-kubernetes==7.5.0
(Didn't see any fix that should change anything between 7.5.0 and 7.6.0 and the code I pointed to has not changed)

### Deployment

Official Apache Airflow Helm Chart

### Deployment details

k8s version: v1.26.3

### Anything else

The issue should be 100% reproductible provided I didn't miss any specifics.

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by Owen-CH-Leung**
fixes #34482 

kubelet client will not remove pods for us. After discussions, it is discovered that all containers inside a Pod will be able to respond to a SIGTERM signal. So instead of going into the istio-proxy container and use the `/quitquitquit` endpoint to pause the sidecar, we can just use the api `delete_namespaced_pod` to delete the entire pod.

This PR therefore changes the logic to delete the pod directly regardless of whether istio is enabled

### CODE REVIEW

**Comment by Owen-CH-Leung** [comment]
Please don't merge this PR - my friend joshua has spotted some additional areas of improvement and will commit to this PR soon

**Comment by hussein-awala** [comment]
> Please don't merge this PR - my friend joshua has spotted some additional areas of improvement and will commit to this PR soon

You should keep your PR as a draft until it is ready

**Comment by Owen-CH-Leung** [comment]
> > Please don't merge this PR - my friend joshua has spotted some additional areas of improvement and will commit to this PR soon
> 
> You should keep your PR as a draft until it is ready

Yup - sorry for that

**Comment by joshuayeung** [comment]
If `istio-proxy` is not running (e.g. the pod is pending and the dag is timeout), `kill_istio_sidecar` will fail.
I added some code to check if istio is running

**Comment by Owen-CH-Leung** [comment]
@hussein-awala Can I seek for your review again ? Many thanks

File: airflow/providers/cncf/kubernetes/operators/pod.py [hunk 1]
```diff
@@ -822,8 +823,12 @@ def process_pod_deletion(self, pod: k8s.V1Pod, *, reraise=True):
...
                     self.pod_manager.delete_pod(pod)
                 elif should_delete_pod and istio_enabled:
-                    self.log.info("Deleting istio-proxy sidecar inside %s: ", pod.metadata.name)
-                    self.kill_istio_sidecar(pod)
+                    if container_is_running(pod, self.ISTIO_CONTAINER_NAME):
+                        self.log.info("Deleting istio-proxy sidecar inside %s: ", pod.metadata.name)
+                        self.kill_istio_sidecar(pod)
+                        self.pod_manager.delete_pod(pod)
+                    else:
+                        self.pod_manager.delete_pod(pod)
```
Comment (jedcunningham) [inline airflow/providers/cncf/kubernetes/operators/pod.py]:
```suggestion
                        self.pod_manager.delete_pod(pod)
```

Doesn't the sidecar listen to sigterm from k8s? In fact this whole thing can be simplified further, if so (no need to see if an istio sidecar is running).

Comment (Owen-CH-Leung) [reply airflow/providers/cncf/kubernetes/operators/pod.py]: hmmm yes. In my istio version I can see that `delete_namespaced_pod` works even when sidecar is running, and seems to me that we don't actually need to `kill_istio_sidecar` at all. However I'm not sure if this behaviour also applies to older version of istio. Shall we keep the logic to end istio sidecar followed by deleting the pod ? Or we should just go ahead to delete the pod directly

Comment (hussein-awala) [reply airflow/providers/cncf/kubernetes/operators/pod.py]: Indeed, when we attempt to terminate the pod, all the containers receive a SIGTERM signal. They manage this through a signal handler. Kubernetes will wait a maximum of `graceperiod` seconds before resorting to a SIGKILL to forcefully terminate them. Given that the KPO has the `termination_grace_period` argument, I concur with Jed's suggestion with a small change; no need for `if should_delete_pod and not istio_enabled` and `if should_delete_pod and istio_enabled`, all this block could be deleted, and we can kill the pod when `should_delete_pod` is Truee regardless the value of `istio_enabled`.

Comment (Owen-CH-Leung) [reply airflow/providers/cncf/kubernetes/operators/pod.py]: Thanks @hussein-awala  and @jedcunningham  for your feedback. I've adopted the suggestion and proceed for deletion whenever `should_delete_pod` is true.

File: tests/providers/cncf/kubernetes/operators/test_pod.py [hunk 2]
```diff
@@ -686,14 +684,6 @@ def test_pod_with_istio_delete_after_await_container_error(
             k.execute(context=context)
 
         assert is_istio_enabled_mock(find_pod_mock.return_value)
```
Comment (jedcunningham) [inline tests/providers/cncf/kubernetes/operators/test_pod.py]: We should either get rid of this test, or test that delete is called. Right now we don't check anything of value any longer, right?

Comment (Owen-CH-Leung) [reply tests/providers/cncf/kubernetes/operators/test_pod.py]: Yes you are right. I've revised the test to check if `delete_pod` is called

**Review by jedcunningham — APPROVED** [review]

**Comment by jedcunningham** [comment]
Can you update the summary/description of this PR now that the approach has changed a bit?

**Comment by Owen-CH-Leung** [comment]
> Can you update the summary/description of this PR now that the approach has changed a bit?

Thanks. I've updated the descriptions

**Review by hussein-awala — APPROVED** [review]

**Pull request merged**